### DISCUSSION TITLE
Error From OpenKeyChain: Missing key for encryption: %s

### DISCUSSION THREAD

**Discussion by TodGilbertHarter** [zeapo/Android-Password-Store#1495] [Q&A] [answered]
I think the title pretty much says it all. Installed OpenKeyChain, created a key pair, set up APS. It seems to be cloning my git repo perfectly well, but I cannot create a password, when I try to save it, I get the above error. 

Also, I'm kind of unclear on how sharing would really work, there wasn't any info on sharing gpg keys between systems. I assume you would have to export both keys of a pair, copy them to the other system, and import them? Granted I don't know if there's really a clean way to do this, but... Certainly some people would probably hope to have that documented ;)

**Comment by msfjarvis** [comment, marked as answer]
OpenKeychain has very poor UX and almost always creates unusable keys. I would recommend creating your GPG key pair on a Linux machine, exporting the secret key, copying it over to your phone and importing it into OpenKeychain, then setting up APS with that key.

**Reply by TodGilbertHarter** [reply]
OK, yeah, I've done that, created a key directly with gpg on my Linux box, exported the private key, copied it to my Android (ugh, could they make that harder) and then imported it into OpenKeychain. THAT WORKS. Cannot agree more on the boo hiss for OKC! I saw you have an issue to replace it. Isn't there just a plain old Bouncy Castle API for this? I mean Java is replete with crypto APIs! If you have to write your own UI elements for it anyway (icky, I know) just go with the best of the best! lol. Trying to JNI in something written in Go has a bad smell, IMHO... (but hey, I'm not a contributor, so definitely my opinion is worth exactly 0 ;)

**Reply by msfjarvis** [reply]
> Cannot agree more on the boo hiss for OKC! I saw you have an issue to replace it. Isn't there just a plain old Bouncy Castle API for this? I mean Java is replete with crypto APIs!

BouncyCastle does have the necessary crypto APIs, but it is a low-level implementation that requires additional wrappers on top to make it more ergonomic to use. It's a lot of hard work which we're both underqualified for and lack the time to do. 

> If you have to write your own UI elements for it anyway (icky, I know) just go with the best of the best! lol. Trying to JNI in something written in Go has a bad smell, IMHO... (but hey, I'm not a contributor, so definitely my opinion is worth exactly 0 ;)

We simply set out to find the most reliably maintained option we had, and Gopenpgp was the one at the time. [PGPainless](https://github.com/pgpainless/pgpainless) emerged later as exactly what we needed, and the option to make use of it is still very much there, provided someone either from the community or one of the maintainers is able to contribute a backend that leverages it.

**Reply by TodGilbertHarter** [reply]
PGPainless does look pretty, well, painless ;) At least for a crypto library. Anyway, I appreciate the software greatly. It looks like it meets my needs. Thanks.
